components/builder-depot-client/src/lib.rs

```diff
@@ -800,13 +802,17 @@ impl Client {
         let result = if let Some(mut progress) = progress {
             progress.size(file_size);
             let mut reader = TeeReader::new(file, progress);
-            self.add_authz(self.0.post_with_custom_url(&path, custom), token)
-                .body(Body::SizedBody(&mut reader, file_size))
-                .send()
+            self.maybe_force_upload(
+                self.add_authz(self.0.post_with_custom_url(&path, custom), token)
+                    .body(Body::SizedBody(&mut reader, file_size)),
+                force_upload,
+            ).send()
         } else {
-            self.add_authz(self.0.post_with_custom_url(&path, custom), token)
-                .body(Body::SizedBody(&mut file, file_size))
-                .send()
+            self.maybe_force_upload(
+                self.add_authz(self.0.post_with_custom_url(&path, custom), token)
+                    .body(Body::SizedBody(&mut file, file_size)),
+                force_upload,
+            ).send()
```

Comment:
As long as we're making changes here, we could first extract out the [`Read`](https://doc.rust-lang.org/nightly/std/io/trait.Read.html) trait in the argument to [`SizedBody`](https://hyper.rs/hyper/0.8.0/hyper/client/enum.Body.html) to reduce duplication and make it more obvious what the difference between the two `add_authz` invocations are:

```rust
        let mut reader: Box<Read> = if let Some(mut progress) = progress {
            progress.size(file_size);
            Box::new(TeeReader::new(file, progress))
        } else {
            Box::new(file)
        };
        let result = self.add_authz(self.0.post(&path), token)
            .body(Body::SizedBody(&mut reader, file_size));
```

Once we do that, we can do away with the `maybe_force_upload` function in favor of the somewhat simpler:

```rust
        header! { (XForce, "X-Force") => [bool] };
        let mut result = self.add_authz(self.0.post(&path), token)
            .body(Body::SizedBody(&mut reader, file_size));
        if force_upload {
            result = result.header(XForce(true));
        }
```
and then match on `result.send()` instead of just `result`.

Or, if it's OK to always send the header, I think this is the simplest I can think of (and `result` doesn't need to be `mut`:

```rust
        let result = self.add_authz(self.0.post(&path), token)
            .body(Body::SizedBody(&mut reader, file_size))
            .header(XForce(force_upload));
```

Reply:
Ok a couple things - this isn't so much a code review as you've fully reimplemented the only logic this PR had in it as well as refactored the original code here. I've got to say, I still have _much_ to learn in rust because your implementation is at the _minimum_ significantly more clever :joy: .

This feels reasonable, the only issue I have with it is that I don't think it makes the code more readable or operable for whoever comes in next, - AND YET it is _definitely_ more dry and I think is likely the right thing. I'm testing this today and will push up changes shortly - thank you for your effort @baumanj. Perhaps we can sync later on what led you to attempt extracting the `Read` trait like you've done? I am a total noob when it comes to traits in rust.

Reply:
We definitely don't want to always send the header.  That should be conditional, based on the presence of the `force` flag.

Reply:
Can you elaborate @raskchanky? What happens in these three scenarios?
1. X-Force: true
2. X-Force: false
3. X-Force not present

Reply:
Perhaps I misspoke.  I haven't looked carefully at @eeyun 's implementation.  In my head, the only reason to send `X-Force` to the server at all is if `--force` has been passed on the command line.  In that case, it would function like item 1 in your list, and the server would re-upload the file.  Items 2 and 3 in your list would be treated the same, namely the behavior that we have today, where we first check to see if we already have the metadata for said file, and do nothing if so.

Reply:
Yeah how I have this implemented on my local branch at this point is that we always send the XForce header with a value of true or false and on the depot server I'm checking the value of the header to determine the behavior. Do we all agree that either method is fine, or should the header only get sent when the value is true?

Reply:
And for further context - I haven't finished that yet so i have no idea if that will actually work.

Reply:
@eeyun That method is fine, provided the server still does the right thing if the header isn't present.  Like I said in my above comment, items 2 & 3 on @baumanj's list should behave the same on the server.